tests/integration/workflows/test_arxiv_workflow.py

```diff
@@ -958,4 +968,45 @@ def test_match_wf_in_error_goes_in_initial_state(workflow_app):
     es.indices.refresh('holdingpen-hep')
 
     with pytest.raises(WorkflowsError):
-        start('article', record)
+        record_workflow = build_workflow(record)
+        start('article', object_id=record_workflow)
+
+
+def test_workflow_loads_from_source_data(workflow_app):
+    broken_record = {
+        'data': {
+            'titles': [
+                {'title': 42},
+            ],
+            'some_messed_up_field': 'will not even validate',
+            'preprint_date': 'Jessica Jones',
+        },
+        'extra_data': {
+            'source_data': {
+                'extra_data': {},
+                'data': {
+                    '_collections': [
+                        'Literature',
+                    ],
+                    'document_type': [
+                        'article',
+                    ],
+                    'titles': [
+                        {'title': 'A title'},
+                    ],
+                }
+            }
+        }
+    }
+
+    record_workflow = workflow_object_class.create(
+        data_type='hep',
+        **broken_record
+    )
+    eng_uuid = start('article', object_id=record_workflow.id)
+
+    eng = WorkflowEngine.from_uuid(eng_uuid)
+    obj = eng.objects[0]
+
+    assert obj.status != ObjectStatus.ERROR
+    assert '_error_msg' not in obj.extra_data
```

Comment:
For the tests, I would:
* e2e testing the feature (creating a workflow that halts, modifying it, restarting and making sure it restored the data and extra_data)
* integration:
** test that it throws the exception on no source_data
** test that it restores the extra_data and data for the rest (probably you might want a custom wokrflow)
*** For wflw with both extra_data and data
*** For wflw with empty extra_data
*** For worfklow with empty data
*** For wflw with both empty
** test that it does not destroy the task_history (check also if all workflows have task_history)